File: together-sandbox-typescript/src/docker.ts
import { promises as fs } from "fs";
import path from "path";
import { writeFile, mkdtemp, rm } from "fs/promises";
import { spawn } from "child_process";
import { tmpdir } from "os";

export async function isDockerAvailable(): Promise<boolean> {
  return new Promise((resolve) => {
    const process = spawn("docker", ["--version"]);
    process.on("close", (code) => {
      resolve(code === 0);
    });
    process.on("error", () => {
      resolve(false);
    });
  });
}

export async function findDockerfile(
  templateDirectory: string,
): Promise<{ exists: boolean; path: string | null }> {
  const dockerfilePath = path.join(templateDirectory, "Dockerfile");
  try {
    await fs.access(dockerfilePath);
    return { exists: true, path: dockerfilePath };
  } catch {
    return { exists: false, path: null };
  }
}

export async function createImageDockerfile(
  image: string,
): Promise<{ dockerfilePath: string; tmpDir: string }> {
  // Create a temporary directory for the Dockerfile
  const tmpDir = await mkdtemp(path.join(tmpdir(), "csb-docker-"));
  const tmpDockerfilePath = path.join(tmpDir, "Dockerfile");

  const dockerfileContent = `FROM ${image}\n`;

  await writeFile(tmpDockerfilePath, dockerfileContent);

  return { dockerfilePath: tmpDockerfilePath, tmpDir };
}

export async function createTemporaryDockerfile(): Promise<string> {
  // Create a temporary directory for the Dockerfile
  const result = await createImageDockerfile("node:24");
  return result.dockerfilePath;
}

export type DockerBuildOptions = {
  dockerfilePath: string;
  imageName: string;
  context: string;
  architecture?: string;
  onOutput?: (output: string) => void;
};

export async function buildDockerImage(
  options: DockerBuildOptions,
): Promise<void> {
  const {
    dockerfilePath,
    imageName,
    context,
    architecture = "amd64",
    onOutput = () => {},
  } = options;

  await new Promise<void>((resolve, reject) => {
    const dockerProcess = spawn(
      "docker",
      [
        "build",
        "--platform",
        `linux/${architecture}`,
        "--progress=plain",
        "-f",
        dockerfilePath,
        "-t",
        imageName,
        context,
      ],
      {
        stdio: ["inherit", "pipe", "pipe"],
      },
    );

    let outputBuffer = "";

    dockerProcess.stdout?.on("data", (data) => {
      const output = data.toString();
      outputBuffer += output;
      // Update spinner with latest output line
      const lines = output.trim().split("\n");
      const lastLine = lines[lines.length - 1];
      if (lastLine) {
        onOutput(lastLine);
      }
    });

    dockerProcess.stderr?.on("data", (data) => {
      const output = data.toString();
      outputBuffer += output;
      // Print stderr to console as well
      const lines = output.trim().split("\n");
      const lastLine = lines[lines.length - 1];
      if (lastLine) {
        onOutput(lastLine);
      }
    });

    dockerProcess.on("close", (code) => {
      if (code === 0) {
        onOutput(`Docker image built successfully: ${imageName}`);
        resolve();
      } else {
        reject(
          new Error(
            `Docker build failed with exit code ${code}\n${outputBuffer}`,
          ),
        );
      }
    });

    dockerProcess.on("error", (error) => {
      reject(`Docker build failed: ${error.message}`);
    });
  });
}

export type DockerLoginOptions = {
  registry?: string;
  username: string;
  password: string;
  onOutput?: (output: string) => void;
};

export async function dockerLogin(options: DockerLoginOptions): Promise<void> {
  const { registry, username, password, onOutput = () => {} } = options;

  await new Promise<void>((resolve, reject) => {
    const args = ["login"];

    if (registry) {
      args.push(registry);
    }

    args.push("--username", username, "--password-stdin");

    const loginProcess = spawn("docker", args, {
      stdio: ["pipe", "pipe", "pipe"],
    });

    // Write password to stdin
    loginProcess.stdin?.write(password);
    loginProcess.stdin?.end();

    let outputBuffer = "";

    loginProcess.stdout?.on("data", (data) => {
      const output = data.toString();
      outputBuffer += output;
      const lines = output.trim().split("\n");
      const lastLine = lines[lines.length - 1];
      if (lastLine) {
        onOutput(lastLine);
      }
    });

    loginProcess.stderr?.on("data", (data) => {
      const output = data.toString();
      outputBuffer += output;
      const lines = output.trim().split("\n");
      const lastLine = lines[lines.length - 1];
      if (lastLine) {
        onOutput(lastLine);
      }
    });

    loginProcess.on("close", (code) => {
      if (code === 0) {
        onOutput(`Docker login successful${registry ? ` to ${registry}` : ""}`);
        resolve();
      } else {
        reject(
          new Error(
            `Docker login failed with exit code ${code}\n${outputBuffer}`,
          ),
        );
      }
    });

    loginProcess.on("error", (error) => {
      reject(new Error(`Docker login failed: ${error.message}`));
    });
  });
}

export async function pushDockerImage(
  imageName: string,
  onOutput?: (output: string) => void,
): Promise<void> {
  onOutput = onOutput || (() => {});

  await new Promise<void>((resolve, reject) => {
    const pushProcess = spawn("docker", ["push", imageName], {
      stdio: ["inherit", "pipe", "pipe"],
    });

    let outputBuffer = "";

    pushProcess.stdout?.on("data", (data) => {
      const output = data.toString();
      outputBuffer += output;
      // Update spinner with latest output line
      const lines = output.trim().split("\n");
      const lastLine = lines[lines.length - 1];
      if (lastLine) {
        onOutput(lastLine);
      }
    });

    pushProcess.stderr?.on("data", (data) => {
      const output = data.toString();
      outputBuffer += output;
      // Print stderr to console as well
      const lines = output.trim().split("\n");
      const lastLine = lines[lines.length - 1];
      if (lastLine) {
        onOutput(lastLine);
      }
    });

    pushProcess.on("close", (code) => {
      if (code === 0) {
        onOutput(`Docker image pushed successfully: ${imageName}`);
        resolve();
      } else {
        reject(
          new Error(
            `Docker push failed with exit code ${code}\n${outputBuffer}`,
          ),
        );
      }
    });

    pushProcess.on("error", (error) => {
      reject(`Docker push failed: ${error.message}`);
    });
  });
}

/**
 * Prepares the Docker build environment for building a Docker image for a Together Sandbox snapshot.
 *
 * @param directory Directory where together-sandbox build is called on
 * @param onOutput Optional output callback for logging
 * @returns A cleanup function to remove temporary Dockerfile if created
 */
export async function prepareDockerBuild(
  directory: string,
  onOutput?: (output: string) => void,
): Promise<{ dockerfilePath: string; cleanupFn: () => Promise<void> }> {
  onOutput = onOutput || (() => {});

  const dockerAvailable = await isDockerAvailable();
  if (!dockerAvailable) {
    console.error(
      "Docker is not available. Please install Docker to use beta build mode.",
    );
    process.exit(1);
  }

  onOutput("Checking for Dockerfile...");

  const dockerfileInfo = await findDockerfile(directory);
  let dockerfilePath: string;
  let needsCleanup = false;
  let tmpDockerfilePath: string | null = null;

  if (!dockerfileInfo.exists) {
    onOutput("Creating temporary Dockerfile...");
    tmpDockerfilePath = await createTemporaryDockerfile();
    dockerfilePath = tmpDockerfilePath;
    needsCleanup = true;
  } else {
    dockerfilePath = dockerfileInfo.path!;
  }

  return {
    dockerfilePath,
    cleanupFn: async () => {
      if (needsCleanup && tmpDockerfilePath) {
        const tmpDir = path.dirname(tmpDockerfilePath);
        await rm(tmpDir, { recursive: true, force: true });
      }
    },
  };
}

File: together-sandbox-python/together_sandbox/docker.py
from __future__ import annotations

import asyncio
import os
import shutil
import tempfile
from collections.abc import Callable
from dataclasses import dataclass


async def is_docker_available() -> bool:
    """Check if Docker is available on the system."""
    try:
        proc = await asyncio.create_subprocess_exec(
            "docker", "--version",
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
        )
        await proc.wait()
        return proc.returncode == 0
    except (FileNotFoundError, OSError):
        return False


async def find_dockerfile(directory: str) -> dict[str, bool | str | None]:
    """
    Check if a Dockerfile exists in the given directory.

    Returns:
        dict with 'exists' (bool) and 'path' (str | None)
    """
    dockerfile_path = os.path.join(directory, "Dockerfile")
    if os.path.exists(dockerfile_path):
        return {"exists": True, "path": dockerfile_path}
    return {"exists": False, "path": None}


async def create_image_dockerfile(image: str) -> dict[str, str]:
    """
    Create a temporary Dockerfile that uses the given image as its base.

    Returns:
        dict with 'dockerfile_path' and 'tmp_dir'
    """
    tmp_dir = tempfile.mkdtemp(prefix="csb-docker-")
    dockerfile_path = os.path.join(tmp_dir, "Dockerfile")
    with open(dockerfile_path, "w") as f:
        f.write(f"FROM {image}\n")
    return {"dockerfile_path": dockerfile_path, "tmp_dir": tmp_dir}


@dataclass
class DockerBuildOptions:
    dockerfile_path: str | None
    image_name: str
    context: str
    architecture: str = "amd64"
    on_output: Callable[[str], None] | None = None


async def _read_stream_lines(
    stream: asyncio.StreamReader,
    on_output: Callable[[str], None],
) -> list[str]:
    lines: list[str] = []
    async for raw in stream:
        line = raw.decode(errors="replace").rstrip("\n")
        lines.append(line)
        if line:
            on_output(line)
    return lines


async def build_docker_image(options: DockerBuildOptions) -> None:
    """Build a Docker image from a Dockerfile."""
    on_output = options.on_output or (lambda _: None)

    proc = await asyncio.create_subprocess_exec(
        "docker", "build",
        "--platform", f"linux/{options.architecture}",
        "--progress=plain",
        *(['-f', options.dockerfile_path] if options.dockerfile_path else []),
        "-t", options.image_name,
        options.context,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    stdout_lines, stderr_lines = await asyncio.gather(
        _read_stream_lines(proc.stdout, on_output),
        _read_stream_lines(proc.stderr, on_output),
    )
    await proc.wait()

    if proc.returncode != 0:
        raise RuntimeError(
            f"Docker build failed with exit code {proc.returncode}\n"
            + "\n".join(stdout_lines + stderr_lines)
        )

    on_output(f"Docker image built successfully: {options.image_name}")


@dataclass
class DockerLoginOptions:
    username: str
    password: str
    registry: str | None = None
    on_output: Callable[[str], None] | None = None


async def docker_login(options: DockerLoginOptions) -> None:
    """Authenticate with a Docker registry via stdin password."""
    on_output = options.on_output or (lambda _: None)

    args = ["docker", "login"]
    if options.registry:
        args.append(options.registry)
    args.extend(["--username", options.username, "--password-stdin"])

    proc = await asyncio.create_subprocess_exec(
        *args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    stdout_data, stderr_data = await proc.communicate(input=options.password.encode())

    for line in stdout_data.decode(errors="replace").splitlines():
        if line:
            on_output(line)
    for line in stderr_data.decode(errors="replace").splitlines():
        if line:
            on_output(line)

    if proc.returncode != 0:
        output = stdout_data.decode(errors="replace") + stderr_data.decode(errors="replace")
        raise RuntimeError(
            f"Docker login failed with exit code {proc.returncode}\n{output}"
        )

    registry_label = f" to {options.registry}" if options.registry else ""
    on_output(f"Docker login successful{registry_label}")


async def push_docker_image(
    image_name: str,
    on_output: Callable[[str], None] | None = None,
) -> None:
    """Push a Docker image to a registry."""
    _on_output = on_output or (lambda _: None)

    proc = await asyncio.create_subprocess_exec(
        "docker", "push", image_name,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    stdout_lines, stderr_lines = await asyncio.gather(
        _read_stream_lines(proc.stdout, _on_output),
        _read_stream_lines(proc.stderr, _on_output),
    )
    await proc.wait()

    if proc.returncode != 0:
        raise RuntimeError(
            f"Docker push failed with exit code {proc.returncode}\n"
            + "\n".join(stdout_lines + stderr_lines)
        )

    _on_output(f"Docker image pushed successfully: {image_name}")